### PR TITLE
core: refactor block_array handling

### PR DESCRIPTION
**Pull request by GBuella**
Introduce routines allocate_new_block_metadata and
unlink_and_release_block_metadata.

---
This change is [Reviewable](https://reviewable.io/reviews/marcinslusarz/pmemfile/33)

### CODE REVIEW

File: src/libpmemfile-core/block_array.c [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+	ASSERT(binfo->idx > 0);
+
+	return binfo->arr->blocks + (binfo->idx - 1);
+}
+
+static void
+unlink_block(struct pmemfile_block *block)
+{
+	if (!TOID_IS_NULL(block->prev))
+		TX_SET_DIRECT(D_RW(block->prev), next, block->next);
```
Comment (marcinslusarz) [inline src/libpmemfile-core/block_array.c]: TX_SET(block->prev, next, block->next)

File: src/libpmemfile-core/block_array.c [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+	if (!TOID_IS_NULL(block->next))
+		TX_SET_DIRECT(D_RW(block->next), prev, block->prev);
```
Comment (marcinslusarz) [inline src/libpmemfile-core/block_array.c]: ditto

File: src/libpmemfile-core/block_array.c [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+static void
+relocate_block(struct pmemfile_block *dst, struct pmemfile_block *src)
+{
+	ASSERT(dst != src);
+
+	TX_ADD_DIRECT(dst);
+
+	if (!TOID_IS_NULL(src->prev))
+		TX_SET_DIRECT(D_RW(src->prev), next, blockp_as_oid(dst));
```
Comment (marcinslusarz) [inline src/libpmemfile-core/block_array.c]: .

File: src/libpmemfile-core/block_array.c [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+	if (!TOID_IS_NULL(src->next))
+		TX_SET_DIRECT(D_RW(src->next), prev, blockp_as_oid(dst));
```
Comment (marcinslusarz) [inline src/libpmemfile-core/block_array.c]: .

File: src/libpmemfile-core/block_array.c [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	update_binfo(vinode);
+
+	if (!has_free_block_entry(vinode))
+		allocate_new_block_array(vinode);
+
+	struct pmemfile_block *block = acquire_new_entry(vinode);
+
+	if (prev == NULL) {
+		if (vinode->first_block != NULL) {
+			block->next = blockp_as_oid(vinode->first_block);
```
Comment (marcinslusarz) [inline src/libpmemfile-core/block_array.c]: missing snapshot

We should probably move snapshotting of the whole block from file_allocate_block_data to acquire_new_entry.

File: src/libpmemfile-core/block_array.c [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+			TX_SET_DIRECT(vinode->first_block,
+			    prev, blockp_as_oid(block));
```
Comment (marcinslusarz) [inline src/libpmemfile-core/block_array.c]: I think we also need to set block->prev to OID_NULL.

File: src/libpmemfile-core/block_array.c [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+		}
+		vinode->first_block = block;
+	} else {
+		block->prev = blockp_as_oid(prev);
+		block->next = prev->next;
```
Comment (marcinslusarz) [inline src/libpmemfile-core/block_array.c]: 2 more missing snapshots

File: src/libpmemfile-core/block_array.c [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "layout.h"
+#include "inode.h"
+#include "internal.h"
+#include "out.h"
+
+static void
+update_binfo(struct pmemfile_vinode *vinode)
```
Comment (marcinslusarz) [inline src/libpmemfile-core/block_array.c]: update_first_block_info?

Comment (GBuella) [inline src/libpmemfile-core/block_array.c]: Done

File: src/libpmemfile-core/block_array.c (hunk 4 above)
Comment (GBuella) [inline src/libpmemfile-core/block_array.c]: Done

File: src/libpmemfile-core/block_array.c (hunk 3 above)
Comment (GBuella) [inline src/libpmemfile-core/block_array.c]: Done

File: src/libpmemfile-core/block_array.c (hunk 2 above)
Comment (GBuella) [inline src/libpmemfile-core/block_array.c]: Done

File: src/libpmemfile-core/block_array.c (hunk 1 above)
Comment (GBuella) [inline src/libpmemfile-core/block_array.c]: Done

File: src/libpmemfile-core/block_array.c (hunk 7 above)
Comment (GBuella) [inline src/libpmemfile-core/block_array.c]: Done

File: src/libpmemfile-core/block_array.c (hunk 5 above)
Comment (GBuella) [inline src/libpmemfile-core/block_array.c]: Done

File: src/libpmemfile-core/block_array.c (hunk 6 above)
Comment (GBuella) [inline src/libpmemfile-core/block_array.c]: The whole new block metadata is expected to be zero, isn't it?

**Comment by marcinslusarz** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/libpmemfile-core/block_array.c, line 121 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/33#-KfBBiYgJ0lA7mPTvxBf-r1-121:-KfBxtFRLPDqfEQVLu-h:ba86dz7) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/63dbf5d66d83fec6a325f5865f5b54e515929271/src/libpmemfile-core/block_array.c#L121)):*
<details><summary><i>Previously, GBuella (Gábor Buella) wrote…</i></summary><blockquote>

Done
</blockquote></details>

"move", not "copy". You need to remove the relevant code from file_allocate_block_data. Please use 2 pmemobj_tx_add_range_direct's like we have in file_allocate_block_data.

---

*[src/libpmemfile-core/block_array.c, line 123 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/33#-KfBBiYgJ0lA7mPTvxBf-r1-123:-KfC-MAC6FRG0PUQk09T:b62eqdp) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/63dbf5d66d83fec6a325f5865f5b54e515929271/src/libpmemfile-core/block_array.c#L123)):*
<details><summary><i>Previously, GBuella (Gábor Buella) wrote…</i></summary><blockquote>

The whole new block metadata is expected to be zero, isn't it?
</blockquote></details>

It's only zeroed becacuse of TX_MEMSET(..., 0,...) in block_list_remove, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/33)*

**Comment by GBuella** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/libpmemfile-core/block_array.c, line 121 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/33#-KfBBiYgJ0lA7mPTvxBf-r1-121:-KfGUYGRKvDd-ZotHxfA:b-896fix) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/63dbf5d66d83fec6a325f5865f5b54e515929271/src/libpmemfile-core/block_array.c#L121)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

"move", not "copy". You need to remove the relevant code from file_allocate_block_data. Please use 2 pmemobj_tx_add_range_direct's like we have in file_allocate_block_data.
</blockquote></details>

Done.

---

*[src/libpmemfile-core/block_array.c, line 123 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/33#-KfBBiYgJ0lA7mPTvxBf-r1-123:-KfGUov8JpRELiKrz5Qc:b2p7wye) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/63dbf5d66d83fec6a325f5865f5b54e515929271/src/libpmemfile-core/block_array.c#L123)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

It's only zeroed becacuse of TX_MEMSET(..., 0,...) in block_list_remove, right?
</blockquote></details>

Also, they are allocated using TX_ZALLOC, I didn't change that part.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/33)*

**Comment by marcinslusarz** [comment]
*[src/libpmemfile-core/block_array.c, line 123 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/33#-KfBBiYgJ0lA7mPTvxBf-r1-123:-KfGZys6Aw5N6N-wM9NH:b-m56feu) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/63dbf5d66d83fec6a325f5865f5b54e515929271/src/libpmemfile-core/block_array.c#L123)):*
<details><summary><i>Previously, GBuella (Gábor Buella) wrote…</i></summary><blockquote>

Also, they are allocated using TX_ZALLOC, I didn't change that part.
</blockquote></details>

With ASSERT(is_zeroed()) in acquire_new_entry I'd feel more comfortable, but it can be done as a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/33)*

**Pull request merged**